[Analysis] Add intra-procedural EscapeAnalysis

apaznikov opened the pull request:
## Summary

This PR introduces a new, generic intra-procedural `EscapeAnalysis` pass. The
analysis determines whether a memory allocation — stack or heap — is confined
to the current function's execution scope, i.e., its address cannot become
visible to other threads.

The pass is designed as a reusable compiler analysis. Its primary initial
motivation and first user is ThreadSanitizer (TSan), where it eliminates
redundant instrumentation for objects that are provably thread-local.
Follow-up TSan integration PR: #192945

Potential future users include alias analysis, scalar optimizations, and other
analyses or transformations that benefit from thread-locality information.

## Implementation Details

The analysis operates intra-procedurally and heavily leverages **MemorySSA**:

1. **MemorySSA-driven dataflow:** The analysis walks the MemorySSA def-use
   chain to trace the lifecycle of pointers through memory operations.
   This allows it to:
   - *Look through loads:* identify the underlying allocation base object
     even if the pointer is loaded from stack slots or aggregates.
   - *Track indirect escapes:* detect scenarios where an address is stored
     into a container object (e.g., a struct or array) and verify whether
     that container allows the pointer to leak via a subsequent load or
     escape of the container itself.

2. **Escape criteria:** An object is considered escaping if:
   - it is stored to a global variable or an escaping object;
   - it is returned from the function;
   - it is passed as an argument to a call that captures the pointer
     (checked via `CaptureTracking`);
   - it is involved in `ptrtoint` casts or volatile operations.

3. **Heap support:** Unlike simple stack-only analyses, this implementation
   identifies non-escaping heap allocations — e.g., a buffer allocated via
   `malloc` and used entirely within the same function.

## Attribution

The underlying algorithms and performance validation were conducted by:
**Alexey Paznikov**, **Michael Schwarz**, **Andrey Kogutenko**, **Yaroslav Osipov**, and **Umang Mathur**.

[1] Preprint: https://arxiv.org/abs/2512.05555

apaznikov committed:
[Analysis] Add intra-procedural EscapeAnalysis

This adds a new generic intraprocedural EscapeAnalysis pass. The analysis determines whether a memory allocation — stack or heap — is confined to the current function's execution scope, i.e., its address cannot become visible to other threads.

The pass is designed as a reusable compiler analysis. Its primary initial motivation and first user is ThreadSanitizer (TSan), where it eliminates redundant instrumentation for objects that areprovably thread-local. Follow-up TSan integration PR: #192945

Potential future users include alias analysis, scalar optimizations, and other analyses that benefit from thread-locality information.

The underlying algorithms and performance validation were conducted by:
Alexey Paznikov, Michael Schwarz, Andrey Kogutenko, Yaroslav Osipov, and Umang Mathur. See https://arxiv.org/abs/2512.05555

apaznikov committed:
[TSan] Add escape analysis for redundant instrumentation elimination

This patch introduces a new intra-procedural Escape Analysis pass and integrates it into ThreadSanitizer to eliminate redundant instrumentation of memory accesses.

The optimization identifies memory allocations (both stack and heap) that are provably thread-local. If an object does not escape the function scope (i.e., is not stored to a global, returned, or passed to a capturing function), it cannot be accessed by another thread, making data races impossible. Consequently, runtime checks for such objects can be safely omitted.

The analysis relies on MemorySSA for precise dataflow tracking. Key features include:
1. Heap & Stack Support: Handles both `alloca` and standard heap allocations (`malloc`, `calloc`, `new`, etc.) via TargetLibraryInfo.
2. MemorySSA-based Tracking: Uses MemorySSA to trace pointer flow through loads and stores. This enables:
   - "Looking through" intermediate memory operations to correctly identify the underlying allocation.
   - Detecting indirect escapes, where a pointer is stored into another local object which is subsequently read or leaked.
3. Conservative Safety: Bails out on unknown flow, complex pointer arithmetic, or when the step limit is exceeded to ensure soundness.

To ensure correctness:
- The analysis is strictly intra-procedural in this patch.
- Volatile accesses and atomic operations are never optimized out.
- It respects TSan's requirements by checking if the base object (not just the pointer) escapes.

This optimization is disabled by default and can be enabled via `-tsan-use-escape-analysis`.

This implementation is based on the research algorithm designed by:
Alexey Paznikov, Andrey Kogutenko, Yaroslav Osipov, Michael Schwarz, and Umang Mathur.

melver commented on llvm/include/llvm/Transforms/Instrumentation/EscapeAnalysis.h:
```diff
@@ -0,0 +1,191 @@
+//===- EscapeAnalysis.h - Intraprocedural Escape Analysis -------*- C++ -*-===//
```
Is this an `Instrumentation` or an `Analysis`? If the latter, it should probably live in `llvm/include/llvm/Analysis/`.

apaznikov replied:
Agreed, EscapeAnalysis is a general-purpose analysis, not instrumentation infrastructure. Moved to `llvm/include/llvm/Analysis/ and llvm/lib/Analysis/`.

apaznikov committed:
[TSan] Move EscapeAnalysis to lib/Analysis and enable by default

Address reviewer comments and prepare the pass for upstreaming:
* Move EscapeAnalysis from Transforms/Instrumentation to llvm/Analysis, as it is a pure analysis pass.
* Enable the optimization by default (-tsan-use-escape-analysis=true).
* Add detailed documentation to the flag explaining safety guarantees.
* Clean up custom lit test configurations as they are no longer needed.
* Include a small drive-by fix for `sanitizer-ld.c` to prevent test failures when the build directory path contains the substring "tsan".

vitalybuka commented on llvm/include/llvm/Analysis/EscapeAnalysis.h:
Also there is StackSafetyAnalisys which does mimilar things for Asan
We had plans to expand that for tsan, but never get to that.

Have you considered expanding functionality of that one?

apaznikov replied:
Thank you for the suggestion. We looked at StackSafetyAnalysis.

The two analyses answer different questions and rely on different mechanisms:
* **StackSafetyAnalysis** proves spatial safety — all accesses to an alloca stay in-bounds.
* **EscapeAnalysis** proves thread locality — a pointer to an allocation cannot be observed by another thread.

These properties are orthogonal: an object can be stack-safe yet have its address stored into a global (causing a race), or be thread-local yet have out-of-bounds accesses. Merging the two would complicate both, since they have no shared algorithmic core. 

Two additional practical differences: EscapeAnalysis handles heap allocations in addition to stack ones, and it works intraprocedurally without requiring ModuleSummaryIndex or ThinLTO. That said, in the future EscapeAnalysis could potentially reuse StackSafetyAnalysis's interprocedural summary infrastructure to extend its reach across call boundaries

vitalybuka replied:
@fmayer Actually I am thinking about removing ThinLTO from StackSafetyAnalisys.
It makes it 2x as complex than it's needed.

Idea was that we can use ThinLTO for better results, but we endup applying all sanitizer before ThinLTO idexing. We endup with this as temporarily solution when switching to new pass manager, but I believe it's good enough as is, and we should not move sanitizers into ThinLTO backends (late sanitizers bad for detection quality).

So ThinLTO parts of StackSafetyAnalisys are dead code, mostly.

I am not sure about MTE. How important it there?
Do MTE users rely on ThinLTO? @fmayer

apaznikov committed:
[Analysis] Decouple EscapeAnalysis from ThreadSanitizer; promote to standalone pass

EscapeAnalysis was previously tightly coupled to ThreadSanitizer.
This commit promotes EscapeAnalysis to a self-contained, consumer-agnostic
pass. The commit also includes additional improvements to the analysis
itself and updates of regression tests.

apaznikov commented:
Gentle ping. @vitalybuka, @dvyukov , @melver — I have addressed the previous feedback by extracting the Escape Analysis into this standalone patch and moving it to llvm/Analysis. The TSan-specific integration will be moved to a follow-up patch to keep this one focused on the analysis itself.

vitalybuka commented:
The description looks very irrelevant, as it focus on Tsan.
It should mention that the first user is Tsan, but I assume it's general analisys, other users are possible.
I suggest just to replace with link to Tsan PR

vitalybuka commented:
@nikic As this looks like a generic thing and you had strong opinion on StackSafetyAnalisys :)

apaznikov commented:
> The description looks very irrelevant, as it focus on Tsan. It should mention that the first user is Tsan, but I assume it's general analisys, other users are possible. I suggest just to replace with link to Tsan PR

@vitalybuka Thanks! I have completely updated the PR description to reflect that this is a generic analysis. I linked the TSan PR as the intended first consumer.

apaznikov commented:
@alinas @aeubanks @fhahn — Hi! Since the primary reviewers are currently overloaded, could one of you kindly take a look at this pass? It is a generic, intra-procedural Escape Analysis heavily relying on MemorySSA. It's fully self-contained, documented, and intended to unblock a 10-15% performance optimization for ThreadSanitizer. Thank you!

nikic commented on llvm/lib/Analysis/EscapeAnalysis.cpp:
```diff
@@ -0,0 +1,818 @@
+
+  // Fallback: isAllocationFn() relies on the allockind attribute being present.
+  // Hand-crafted IR or old bitcode may lack it even for standard functions.
+  // This TLI-direct check catches those cases.
```
Just a small drive-by note while scrolling through: This function should not exist. You should assume that allocation functions are appropriately annotated. There is no need to handle non-canonical IR.

apaznikov replied:
Done, thanks! I removed the fallback and now rely on the allocation function annotations as suggested.
